lib/index.js

```diff
@@ -669,17 +653,35 @@ Socket.prototype._flushReads = function() {
 
   this._isFlushingReads = true;
 
-  var received;
+  var message;
 
   do {
     try {
-      received = this._flushRead();
+      message = this._zmq.readv(); // can throw
     } catch (error) {
       this._isFlushingReads = false;
       this.emit('error', error); // can throw
       return;
     }
-  } while (received);
+
+    if (message) {
+      try {
+        // Handle received message immediately to prevent memory leak in driver
+        if (message.length === 1) {
+          // hot path
+          this.emit('message', message[0]);
+        } else {
+          this.emit.apply(this, ['message'].concat(message));
+        }
+      } catch (error) {
+        // There might be additional unprocessed messages so continue reading
+        // after we throw the exception.
+        process.nextTick(this._flushReads.bind(this));
```

Comment:
We've been here in the past, and it was the source of all kinds of trouble iirc. Allow me some time to review this in more depth.